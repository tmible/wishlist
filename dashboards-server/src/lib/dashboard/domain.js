import { PERIOD } from '$lib/constants/period.const.js';

/** @module Домен дашборда */

/**
 * Тип данных дашборда
 * @template T
 * @typedef {{ [chart: string]: undefined | null | T }} DashboardData
 */
/**
 * Базовый конфиг для построения дашборда
 * @template T
 * @typedef {object} DashboardConfigBase
 * @property {'line' | 'doughnut'} type Тип графиков дашборда
 * @property {PERIOD} [period] Период для отображения графиков
 * @property {DashboardData<T>} [initialData] Данные для инициализации дашборда
 */
/** @typedef {{ [label: string]: unknown, value: number }[]} LineDashboardData */
/**
 * Конфиг для построения дашборда с графиками в виде линий
 * @typedef {object} LineDashboardConfig
 * @property {'line'} type Тип графиков дашборда
 * @property {object} chartConfigs Конфиг грфиков
 * @property {string} chartConfigs.key Идентификатор графика
 * @property {string} chartConfigs.label Название графика для отображения
 * @property {boolean} [chartConfigs.isDisplayed] Признак отображения графика
 * @augments DashboardConfigBase<LineDashboardData>
 */
// eslint-disable-next-line no-secrets/no-secrets -- Название типа с шаблоном
/**
 * Конфиг для построения дашборда с графиком в виде кольца
 * @typedef {object} DoughnutDashboardConfig
 * @property {'doughnut'} type Тип графика дашборда
 * @property {string} key Идентификатор графика
 * @property {string} label Название графика для отображения
 * @augments DashboardConfigBase<number>
 */
/**
 * Конфиг для построения дашборда
 * @typedef {LineDashboardConfig | DoughnutDashboardConfig} DashboardConfig
 */

/**
 * Базовый дашборд
 * @template T
 * @typedef {object} DashboardBase
 * @property {PERIOD} period Период для отображения графиков
 * @property {DashboardData<T>} [initialData] Данные для инициализации дашборда
 */
/** @typedef {{ [label: string]: unknown, value: number }[]} LineDashboardData */
/**
 * Дашборд с графиками в виде линий
 * @typedef {object} LineDashboard
 * @property {
 *   Map<
 *     LineDashboardConfig['chartConfigs']['key'],
 *     {
 *       label: LineDashboardConfig['chartConfigs']['label'],
 *       isDisplayed: boolean,
 *       data: DashboardData<LineDashboardData>,
 *     }
 *   >
 * } charts Грфики дашборда
 * @augments DashboardBase<LineDashboardData>
 */
/**
 * Дашборд с графиком в виде кольца
 * @typedef {object} DoughnutDashboard
 * @property {
 *   Map<
 *     DoughnutDashboardConfig['key'],
 *     {
 *       label: DoughnutDashboardConfig['label'],
 *       isDisplayed: boolean,
 *       data: DashboardData<number>,
 *     }
 *   >
 * } charts Грфик дашборда
 * @augments DashboardBase<number>
 */
/**
 * Дашборд
 * @typedef {LineDashboard | DoughnutDashboard} Dashboard
 */

/**
 * Период для отображения графиков по умолчанию
 * @constant {PERIOD}
 */
const DEFAULT_PERIOD = PERIOD.DAY;

/**
 * Строители дашбордов с разными типами графиков
 * @constant {{
 *   [chartType: DashboardConfig['type']]: (
 *     dashboard: Dashboard,
 *     dashboardConfig: DashboardConfig,
 *   ) => void
 * }}
 */
const BUILDERS = {
  line: (dashboard, { chartConfigs }) => {
    dashboard.charts = new Map(chartConfigs.map(({ key, label, isDisplayed }, i) => [
      key,
      {
        label,
        isDisplayed: isDisplayed ?? i === 0,
        data: undefined,
      },
    ]));
  },
  doughnut: (dashboard, { key, label }) => {
    dashboard.charts = new Map([[
      key,
      {
        label,
        isDisplayed: true,
        data: undefined,
      },
    ]]);
  },
};

/**
 * Построение дашборда. Установка общих свойств и вызов строителя в соответствии с типом
 * @function buildDashboard
 * @param {DashboardConfig} dashboardConfig Конфиг дашборда
 * @returns {Dashboard} Построенный дашборд
 */
export const buildDashboard = ({
  period = DEFAULT_PERIOD,
  type,
  initialData = {},
  ...restConfig
}) => {
  const dashboard = { period, initialData };
  BUILDERS[type](dashboard, restConfig);
  return dashboard;
};
