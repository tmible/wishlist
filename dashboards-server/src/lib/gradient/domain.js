import numericEnum from '@tmible/wishlist-common/numeric-enum';
import { getRandomArbitrary } from '$lib/get-random-arbitrary.js';

/** @module Домен градиента */

/**
 * Градиент
 * @typedef {object} Gradient
 * @property {number} hue1 Тон первого цвета в градиенте
 * @property {number} hue2 Тон второго цвета в градиенте
 * @property {number} saturation Насыщенность цветов в градиенте
 * @property {number} lightness Светлота цветов в градиенте
 * @property {GradientVariant} variant Вариант градиента
 */

/**
 * Минимальный и максимальный тон цвета в градиенте
 * @constant {[ number, number ]}
 */
const HUE_MIN_MAX = [ 0, 360 ];

/**
 * Минимальная и максимальная разница в тонах цветов в градиенте
 * @constant {[ number, number ]}
 */
const HUE_DIFF_MIN_MAX = [ 30, 60 ];

/**
 * Минимальная и максимальная насыщенность цветов в градиенте
 * @constant {[ number, number ]}
 */
const SATURATION_MIN_MAX = [ 50, 100 ];

/**
 * Минимальная и максимальная светлота цветов в градиенте в тёмном варианте
 * @constant {[ number, number ]}
 */
const LIGHTNESS_DARK_MIN_MAX = [ 10, 33 ];

/**
 * Минимальная и максимальная светлота цветов в градиенте в светлом варианте
 * @constant {[ number, number ]}
 */
const LIGHTNESS_LIGHT_MIN_MAX = [ 67, 90 ];

/**
 * Перечисление вариантов градиента
 * @enum {number}
 */
export const GradientVariant = numericEnum([ 'LIGHT', 'DARK' ]);

/**
 * Генерация градиента
 * @function generateGradient
 * @param {GradientVariant} variant Вариант градиента
 * @returns {Gradient} Сгенерированный градиент
 */
export const generateGradient = (variant = GradientVariant.LIGHT) => {
  const hue1 = getRandomArbitrary(...HUE_MIN_MAX);
  const hue2 = (
    // eslint-disable-next-line sonarjs/pseudo-random -- Просто генерация градиента
    hue1 + ([ -1, 1 ][Math.round(Math.random())] * getRandomArbitrary(...HUE_DIFF_MIN_MAX))
  ) % 360;
  const saturation = getRandomArbitrary(...SATURATION_MIN_MAX);
  const lightness = getRandomArbitrary(
    ...(variant === GradientVariant.LIGHT ? LIGHTNESS_LIGHT_MIN_MAX : LIGHTNESS_DARK_MIN_MAX),
  );

  return { hue1, hue2, saturation, lightness, variant };
};

/**
 * Приведение градиента к нужному варианту или генерация нового при невозможности
 * @function changeGradientVariant
 * @param {Gradient} currentGradient Активный градиент
 * @param {GradientVariant} variant Целевой вариант градиента
 * @returns {Gradient} Адаптированный градиент
 */
export const changeGradientVariant = (currentGradient, variant) => {
  const gradient = currentGradient ?? generateGradient(variant);

  if (gradient.variant !== variant) {
    gradient.lightness = getRandomArbitrary(
      ...(variant === GradientVariant.LIGHT ? LIGHTNESS_LIGHT_MIN_MAX : LIGHTNESS_DARK_MIN_MAX),
    );
    gradient.variant = variant;
  }

  return gradient;
};
