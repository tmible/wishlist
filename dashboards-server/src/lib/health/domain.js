/**
 * Информация о здоровье бота
 * @typedef {object} BotHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} localhost Запущен ли http сервер
 * @property {boolean} https Проксируется ли https запрос к домену на http сервер
 * @property {boolean} dbConnection Подключен ли бот к БД
 * @property {boolean} localDBConnection Подключен ли бот к БД для горячих данных
 * @property {boolean} hubConnection Подключен ли бот к хабу
 */
/**
 * Информация о здоровье портала
 * @typedef {object} PortalHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} localhost Запущен ли http сервер
 * @property {boolean} https Проксируется ли https запрос к домену на http сервер
 * @property {boolean} dbConnection Подключен ли портал к БД
 * @property {boolean} hubConnection Подключен ли портал к хабу
 */
/**
 * Информация о здоровье хаба
 * @typedef {object} HubHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} socket Принимает ли хаб подключения к UNIX доменному сокету
 */
/**
 * Информация о здоровье сервиса удаления протухших refresh токенов аутентификации на портале
 * @typedef {object} RefreshTokensCleanerHealthData
 * @property {boolean} timer Запущен ли systemd таймер
 * @property {boolean} service Запущен ли systemd сервис
 */
/**
 * Информация о проверке здоровья сервисов
 * @typedef {object} HealthData
 * @property {number} date Таймштамп проверки
 * @property {BotHealthData | null} bot Информация о здоровье бота
 * @property {PortalHealthData | null} portal Информация о здоровье портала
 * @property {HubHealthData | null} hub Информация о здоровье хаба
 * @property {RefreshTokensCleanerHealthData | null} refreshTokensCleaner
 * Информация о здоровье сервиса удаления протухших refresh токенов аутентификации на портале
 * @property {object | null} other Объединение всех сервисов, кроме бота и портала
 */

/**
 * Добавление объединения информации о сервисах, кроме бота и портала
 * @function addOtherService
 * @param {HealthData} health Информация о здороье сервисов
 * @returns {HealthData} Информация с добавленным объединением
 */
export const addOtherService = (health) => {
  health.other = {};
  for (const service of Object.entries(health)) {
    if (service[0] === 'bot' || service[0] === 'portal' || service[0] === 'other') {
      continue;
    }
    Object.assign(
      health.other,
      Object.fromEntries(
        Object.entries(service[1]).map(([ key, value ]) => [ `${service[0]}.${key}`, value ]),
      ),
    );
  }
  return health;
};
