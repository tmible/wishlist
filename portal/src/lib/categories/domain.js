/** @module Домен категорий */

/**
 * Категория элементов списка желаний пользователя
 * @typedef {object} Category
 * @property {number} id Идентификатор категории
 * @property {string} name Название категории
 */

/**
 * Проверка наличия различий в двух экземплярах категории.
 * Оба объекта должны быть экземплярами одной и той же категории
 * @function isCategoryEdited
 * @param {Category} category Оригинальный экземпляр категории
 * @param {Category} editedCategory Изменённый экземпляр категории
 * @returns {boolean} Признак наличия различий
 */
export const isCategoryEdited = (
  category,
  editedCategory,
) => category.id === editedCategory.id && category.name !== editedCategory.name;
