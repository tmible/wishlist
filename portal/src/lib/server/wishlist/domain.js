/** @typedef {import('$lib/wishlist/domain.js').OwnWishlistItem} OwnWishlistItem */

/** @module Домен списка желаний */

/**
 * Присовение порядкового номера новому (добавляемому) подарку
 * @function assignOrderToNewItem
 * @param {OwnWishlistItem[]} list Текущий список
 * @param {Partial<OwnWishlistItem>} itemToAdd Добавляемый подарок
 * @returns {void}
 */
export const assignOrderToNewItem = (list, itemToAdd) => {
  itemToAdd.order = list.length;
};

/**
 * Отметка подарка сюрпризом
 * @function markExernallyAddedItem
 * @param {OwnWishlistItem} item Добавляемый подарок
 * @param {number} addedBy Идентификтор пользователя, добавившего подрарок в список желаний
 * @returns {void}
 */
export const markExernallyAddedItem = (item, addedBy) => {
  item.addedBy = addedBy;
};
