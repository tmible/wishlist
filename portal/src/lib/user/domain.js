import { DEFAULT_USER } from './default-user.const.js';

/** @module Домен пользователя */

/**
 * Пользователь
 * @typedef {object} User
 * @property {number | null} id Идентификатор пользователя
 * @property {string | null} hash Хэш пользователя
 * @property {boolean | null} isAuthenticated Признак аутентифицированности пользователя
 */

/**
 * Создание пользователя
 * @function createUser
 * @param {Partial<User>} initialData Инициализационные данные
 * @returns {User} Созданный пользователь
 */
export const createUser = (initialData) => ({
  ...DEFAULT_USER,
  ...initialData,
});

/**
 * Аутентификация пользователя
 * @function loginUser
 * @param {User} user Пользователь
 * @param {User['userid']} userid Идентификатор пользователя
 * @returns {User} Обновлённый пользователь
 */
export const loginUser = (user, userid) => {
  user.isAuthenticated = true;
  user.id = userid;
  return user;
};

/**
 * Разлогинивание пользователя
 * @function logoutUser
 * @param {User} user Пользователь
 * @returns {User} Обновлённый пользователь
 */
export const logoutUser = (user) => {
  user.isAuthenticated = false;
  user.id = null;
  user.hash = null;
  return user;
};

/**
 * Установка хэша пользователя
 * @function setUserHash
 * @param {User} user Пользователь
 * @param {User['hash']} hash Хэш пользователя
 * @returns {User} Обновлённый пользователь
 */
export const setUserHash = (user, hash) => {
  user.hash = hash;
  return user;
};
