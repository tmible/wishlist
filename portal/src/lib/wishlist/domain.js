import arrayToOrderedJSON from '@tmible/wishlist-common/array-to-ordered-json';

/** @typedef {import('@tmible/wishlist-common/constants/list-item-state').default} ListItemState */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/** @typedef {import('$lib/categories/domain.js').Category} Category */

/**
 * Элемент списка желаний пользователя во варианте отображения для владельца
 * @typedef {object} OwnWishlistItem
 * @property {number} id Идентификатор элемента
 * @property {string} name Название подарка
 * @property {string} description Описание подарка
 * @property {ListItemState} state Состояние подарка
 * @property {number} order Порядковый номер элемента
 * @property {Category | null} category Категория элемента
 * @property {Entity[]} descriptionEntities Элементы разметки текста описания подарка
 */

/**
 * Свойства элемента списка, определяемые при его создании или изменении
 * @typedef {
 *   Omit<OwnWishlistItem, 'id' | 'state' | 'order' | 'category'> &
 *     { order?: number, categoryId: number | null }
 * } OwnWishlistItemFormData
 */

/**
 * Присовение порядкового номера новому (создаваемому) элементу списка
 * @function assignOrderToNewItem
 * @param {OwnWishlistItem[]} list Текущий список
 * @param {Partial<OwnWishlistItem>} itemToAdd Создаваемый элемент
 * @returns {void}
 */
export const assignOrderToNewItem = (list, itemToAdd) => {
  itemToAdd.order = list.length;
};

/**
 * Определение наличия изменений, фильтрация объекта с изменениями
 * и создание нового изменённого элемена списка
 * @function patchItem
 * @param {OwnWishlistItem} item Изменяемый элемент списка
 * @param {Partial<OwnWishlistItem>} patch Объект с изменениями
 * @returns {[ boolean, OwnWishlistItem, Partial<OwnWishlistItem> ]}
 *   Признак наличия изменений, изменённый элемент списка и отфильтрованный объект с изменениями
 */
export const patchItem = (item, patch) => {
  const patchFiltered = { ...patch };
  for (const [ key, value ] of Object.entries(patch)) {
    if (
      (
        key === 'descriptionEntities' &&
        arrayToOrderedJSON(value) === arrayToOrderedJSON(item[key])
      ) ||
      (key === 'category' && (value?.id ?? null) === (item.category?.id ?? null)) ||
      (
        key !== 'descriptionEntities' &&
        key !== 'category' &&
        value === item[key]
      )
    ) {
      delete patchFiltered[key];
    }
  }

  return [ Object.entries(patchFiltered).length > 0, { ...item, ...patchFiltered }, patchFiltered ];
};
