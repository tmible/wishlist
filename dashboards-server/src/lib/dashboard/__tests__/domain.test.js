import { describe, expect, it } from 'vitest';
import { PERIOD } from '$lib/constants/period.const.js';
import { buildDashboard } from '../domain.js';

describe('dashboard / domain', () => {
  it('should build dashboard in minimal variant', () => {
    expect(buildDashboard({ type: 'line', chartConfigs: [] })).toMatchSnapshot();
  });

  it('should build dashboard in maximal variant', () => {
    expect(
      buildDashboard({
        type: 'line',
        period: PERIOD.HOUR,
        initialData: { 0: 'initial', 1: 'data' },
        chartConfigs: [{
          key: 0,
          label: '0',
          isDisplayed: false,
        }, {
          key: 1,
          label: '1',
          isDisplayed: true,
        }],
      }),
    ).toMatchSnapshot();
  });

  it('should build line dashboard', () => {
    expect(
      buildDashboard({
        type: 'line',
        chartConfigs: [{
          key: 0,
          label: '0',
        }, {
          key: 1,
          label: '1',
        }],
      }),
    ).toMatchSnapshot();
  });

  it('should build doughnut dashboard', () => {
    expect(buildDashboard({ type: 'doughnut', key: 'key', label: 'label' })).toMatchSnapshot();
  });
});
