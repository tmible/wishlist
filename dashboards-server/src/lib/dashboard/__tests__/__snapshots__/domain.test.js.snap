// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dashboard / domain > should build dashboard in maximal variant 1`] = `
{
  "charts": Map {
    0 => {
      "data": undefined,
      "isDisplayed": false,
      "label": "0",
    },
    1 => {
      "data": undefined,
      "isDisplayed": true,
      "label": "1",
    },
  },
  "initialData": {
    "0": "initial",
    "1": "data",
  },
  "period": 3600000,
}
`;

exports[`dashboard / domain > should build dashboard in minimal variant 1`] = `
{
  "charts": Map {},
  "initialData": {},
  "period": 86400000,
}
`;

exports[`dashboard / domain > should build doughnut dashboard 1`] = `
{
  "charts": Map {
    "key" => {
      "data": undefined,
      "isDisplayed": true,
      "label": "label",
    },
  },
  "initialData": {},
  "period": 86400000,
}
`;

exports[`dashboard / domain > should build line dashboard 1`] = `
{
  "charts": Map {
    0 => {
      "data": undefined,
      "isDisplayed": true,
      "label": "0",
    },
    1 => {
      "data": undefined,
      "isDisplayed": false,
      "label": "1",
    },
  },
  "initialData": {},
  "period": 86400000,
}
`;
